### PR TITLE
ARTEMIS-1573 Improve UTF translation allowing zero copy

### PR DESCRIPTION
**Pull request by franz1981**
The UTF translations has been improved by:
- zero copy on array based buffers
- zero copy UTF length calculation
- faster array access using Netty PlatformDependent.get|putByte
- improved perf tests UTF8Test

### CODE REVIEW

**Comment by franz1981** [comment]
Please do not merge yet: I've already run the CI and compatibility tests (kudos to @clebertsuconic work on it!), but I will need to provide some results on it :+1:

File: artemis-commons/src/main/java/org/apache/activemq/artemis/utils/UTF8Util.java [hunk 1]
```diff
@@ -34,7 +35,7 @@
 
    private static final boolean isTrace = ActiveMQUtilLogger.LOGGER.isTraceEnabled();
 
-   private static final ThreadLocal<SoftReference<StringUtilBuffer>> currenBuffer = new ThreadLocal<>();
+   private static final FastThreadLocal<SoftReference<StringUtilBuffer>> currentBuffer = new FastThreadLocal<>();
```
Comment (michaelandrepearce) [inline artemis-commons/src/main/java/org/apache/activemq/artemis/utils/UTF8Util.java]: This only benefits if all threads are FastThreadLocalThread's. Do we re-use Netty's thread pools? I thought we had our own....

Comment (clebertsuconic) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/UTF8Util.java]: @michaelandrepearce +1

Comment (franz1981) [reply artemis-commons/src/main/java/org/apache/activemq/artemis/utils/UTF8Util.java]: @michaelandrepearce Good catch! I will change it in no time, thanks!

**Comment by michaelandrepearce** [comment]
@franz1981 LGTM. 

The only comment i would have, is it worth adding the code to handle the offheap now also, just while its fresh in your mind, and then it be more complete feature. And if you disagree then thats also fine, id be still happy for this to merge.

**Comment by franz1981** [comment]
@michaelandrepearce 
> is it worth adding the code to handle the offheap now also, just while its fresh in your mind, and then it be more complete feature

Agree :) I've avoided to implement them (for the `writeAsShort` case too) to not turn this PR in a total rewrite of the original code, but I already know that the gains are *huge*: I will take some time to close the circle and make it features complete by covering the remaining cases :+1:

**Comment by franz1981** [comment]
I've changed the last few bits and these are some results:
```
THIS PR:
DIRECT:
Throughput writeUTF = 2352 ops/ms
Throughput readUTF = 2123 ops/ms
HEAP:
Throughput writeUTF = 2531 ops/ms
Throughput readUTF = 2257 ops/ms

ORIGINAL
DIRECT;
Throughput writeUTF = 1824 ops/ms
Throughput readUTF = 1494 ops/ms
HEAP:
Throughput writeUTF = 1828 ops/ms
Throughput readUTF = 1727 ops/ms
```

**Comment by michaelandrepearce** [comment]
@franz1981 looks good, ill look to merge this afternoon, just to confirm no more bits right?

**Pull request closed without merging**